File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void bfs(int row, int col, vector<vector<bool>>& vis, vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        
        queue<pair<int, int>> q;
        q.push({row, col});
        vis[row][col] = true;  // mark visited when pushed
        
        vector<pair<int, int>> dir = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            
            for (auto& d : dir) {
                int nr = r + d.first;
                int nc = c + d.second;
                
                if (nr >= 0 && nr < n &&
                    nc >= 0 && nc < m &&
                    grid[nr][nc] == '1' &&
                    !vis[nr][nc]) {
                    
                    vis[nr][nc] = true;   // mark before pushing
                    q.push({nr, nc});
                }
            }
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        int totalIsland = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!visited[i][j] && grid[i][j] == '1') {
                    bfs(i, j, visited, grid);
                    totalIsland++;
                }
            }
        }
        return totalIsland;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        islands = 0
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        visited = [[False] * cols for _ in range(rows)]
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and not visited[r][c] :
                    islands += 1
                    q = deque([(r, c)])
                    while q:
                        row, col = q.popleft()

                        for dr, dc in directions:
                            nr = row + dr
                            nc = col + dc
                            if (
                                0 <= nr < rows and
                                0 <= nc < cols and
                                grid[nr][nc] == "1" and not visited[nr][nc]

                            ):
                                visited[nr][nc] = True
                                q.append((nr, nc))
        return islands
        
